### DISCUSSION TITLE
[Discussion] Possible to have an option to "snap" location to roads/paths?

### DISCUSSION THREAD

**Discussion by wiseindy** [Freika/dawarich#43] [Ideas]
Firstly, I would like to say that this project is amazing! I've always wanted to store and view my Google timeline data offline and this does the job extremely well.

I've imported all my Google data to DaWarIch and it's great. I've noticed that when we view the same data in Google Timeline as compared to DaWarIch, Google tends to "snap" the location points to roads (when driving, at least) and that results in a much cleaner look instead of straight lines going across the map.

Do we know if something like that is even possible with OpenStreetMap + DaWarIch?

**Comment by wiseindy** [comment]
I think this is related https://github.com/Freika/dawarich/issues/34

**Comment by Freika** [comment]
Snapping sounds like a good idea, although the very first thought on that is: how it should choose between two different roads leading to a certain point from the previous one? It might be pretty straightforward when driving, but on other cases might be tricky.

 But it certainly something I'll think about, thank you!

**Reply by reidprichard** [reply]
I think this would be a pretty cool feature too - not only to snap points to roads, but also to smooth the lines between points.

Looks like [Google uses a series of points](https://developers.google.com/maps/documentation/roads/snap). That way you can eliminate possibilities based on traffic patterns (e.g., you turned across a freeway where there's no off-ramp, so you must have been on the frontage road on the specific side rather than on the freeway). 

It looks like OpenStreetMap has that information encoded into [ways and relations](https://wiki.openstreetmap.org/wiki/Elements). 

While technically possible, I'm not sure how feasible this would be. I imagine it would mean a LOT of OSM API calls, and I could see it being pretty resource-intensive to fit a large number of points to a street map. Not to mention the programming effort!

Just for fun - as someone with zero prior knowledge of anything relevant - I'd guess the process would go something like this:
1. Take a batch of X sequential points
2. Iterate over the points. 
    a. At each point, get a list of roads ("highways" in OSM parlance) within distance "Y". Looks like you can get a list of objects within a rectangular region, so the easy way would be to get the list of highways in the box defined by +/- Y in both axes.
    b. Find the shortest Z paths that connect the closest point of each of those roads to the next point. 
    c. Rule out any paths that aren't possible based on the previous point. For example, Point B couldn't have been reached by a freeway if there wasn't an onramp in between it and Point A.
3. At the end, if no highways remain, increase the tolerance Y and start over from step 1. If more than one highway remains, reduce the tolerance Y and start over from step 1.
Ok, hugely inefficient and I skipped all the hard parts. 2b and 2c seem very challenging, and 3 being iterative makes this like O(N^3) expensive probably. I think you could use a traveling salesman method for 2b, but it seems easier to predict the optimal path than to find the actual one - you don't know that an optimal route was taking, and aliasing errors are inevitable. Not to mention the expense of finding the nearest point along each road. Then in 2c you have to figure out how far you'll go along a road before saying it isn't possible. I suppose you could estimate this based on times and speeds of the GPS data.

I don't know if I contributed anything useful but it was fun to brainstorm!

**Comment by kobius77** [comment]
there is ([or was](https://github.com/paulmach/slide/issues/13 )) [Strava Slide](https://labs.strava.com/slide/). Its been offered as an OSM-editor and worked quite well, but i don't know anything about it technically or legally 🤷

**Comment by kobius77** [comment]
dumb thought: doesn't _any_ navigation tool based on OSM data create a line that snaps onto existing roads? 😗

**Reply by NovaCyntax** [reply]
The "[OsmAnd](https://github.com/osmandapp/OsmAnd)" app (FOSS Google Maps replacement) does technically support generating .gpx files from your trips if I remember correctly. Would be interested to know if the data generated from that is "snapped" onto the roads, or if it's the raw GPS values (like what we are currently using). If it's the former, the application is open-source, and their implementation could be used as a reference.

Another FOSS app with a similar capability is "[Organic Maps](https://git.omaps.dev/organicmaps/organicmaps)". It has a smaller feature-set than OsmAnd, and might be easier to reference. Additionally, the logic for "snapping" as it pertains to .gpx file generation may vary between the two apps.

**Comment by NovaCyntax** [comment]
This would clean up my map a lot, especially when zoomed in for greater detail. I'm not sure how you could prevent false-positives from happening though.

**Comment by Praaug** [comment]
I would also really like this feature.

I'm using dawarich for a week now and am constantly reducing the minimal distance in owntracks to have it somewhat lined up with the roads. But especially on highway on- / offramps or interchanges it can't get precise.

I know a few years ago on goolge timeline it would do the same and you could snap them with a button click.
I think something like this could work here as well. That way you only get it when you want to (driving by car) but not when on the train or going for a walk in the forest, where it can't snap to anything.
That would also decrease the amounts of calls to the osm api in comparison to snapping everything.

As an idea for choosing the right point:
You could ask the navigation api and calculate a route from start to finish, and add every x point as waypoints in the middle. Although I have no Idea how to verify that. Or ask the navigation between two points, where you are more certain they are on a specific road.
but that could get quite complex when the route is not A->B directly

**Comment by baptisteba** [comment]
Worth checking again to include feature as there is 2 great projects that can do it perfectly, the same way we use photon for reverse geocoding : 

- https://github.com/valhalla/valhalla ⭐
- https://project-osrm.org/ _Public server available for testing_

Self Hosting possible for both
